### PR TITLE
feat(api): feedback ingestion endpoint

### LINKED ISSUE #1027: feat(api): feedback ingestion endpoint
Part of epic #1026.

## Context

Foundation brick of the beta feedback loop. Both the website widget (#1026 → C3) and the in-app feedback (#1026 → C4) reuse the `feedback-widget` project, which is frontend-only and POSTs to a configurable endpoint. That endpoint must exist in the centralized NestJS API (ADR-0002).

## Goals

- `POST /feedback` endpoint whose contract matches the payload emitted by `feedback-widget`'s `core` package (category + free-text description + context metadata).
- Persist from day one (no demo-only mode) — schema + migration.
- Attach feedback to the single source of truth for consent (ADR-0003); feedback may contain personal data.
- `projectId` discriminator so the same endpoint can serve site vs app sources.

## Acceptance

- Endpoint validated (DTO), persisted, returns proper status codes.
- Tests: happy + sad + edge cases.
- No `any`, named exports only.

## Dependencies

- Blocks C3 (website widget) and C4 (in-app feedback).

## Reference

- Payload contract: see `feedback-widget` `packages/core` (categories, validation, state machine).
- ADR-0002 (centralized backend), ADR-0003 (consent SSoT).

---

## Scope correction (v0.1) — consent is client-side

Per [ADR-0003](../architecture/decisions/0003-consent-single-source-of-truth.md) and the current API code, **there is no backend consent module at v0.1** — the consent store lives on the mobile client; backend sync is the ADR-0003 v0.2 roadmap. So this endpoint:

- **Does NOT** implement or call a backend consent gate at v0.1. It persists what it receives.
- Is a **public / anonymous** route (website visitors + app users), rate-limited (`@Throttle`) against spam.
- Minimises PII: `reporterName` optional, `message` treated as potentially-PII free text.
- Anonymous-visitor consent on the **website** stays an open question (widget checkbox vs documented legitimate-interest basis) — resolve with #1028.

Server-side consent gating is deferred to the v0.2 backend consent sync. See the corrected UML (#1030 / #1040).

### PR DESCRIPTION
**Pull request by benoit-bremaud**
C2 of epic #1026. Public POST /feedback endpoint (anonymous, rate-limited) persisting feedback to a new table. Mirrors the feedback-widget FeedbackPayload contract; server-side pairing validation. No backend consent gate at v0.1 (ADR-0003, client-side). Closes #1027

### CODE REVIEW

**Comment by benoit-bremaud** [comment]
@vitalikevin — Pour info : voici la première brique technique de la boucle de feedback (chantier #1027). C'est le point d'entrée côté serveur qui reçoit et enregistre les avis envoyés depuis le site et l'app. Route publique (pas besoin de compte), protégée contre le spam, et qui valide le contenu reçu. Pas de gestion du consentement côté serveur pour l'instant : à ce stade c'est géré côté client (ADR-0003), la synchro serveur viendra plus tard. Rien à faire de ton côté dans l'immédiat ; ça débloque l'intégration du widget sur le site (#1028) et dans l'app (#1029).

File: packages/api/src/feedback/feedback.module.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+/**
+ * Feedback Module (epic #1026, C2 / #1027)
+ *
+ * Registers the public `POST /feedback` endpoint, its persistence, and
+ * the sub-category pairing validator. No backend consent dependency at
+ * v0.1 (ADR-0003 keeps consent client-side; backend sync is v0.2).
+ */
+@Module({
+  imports: [TypeOrmModule.forFeature([Feedback])],
```
[the remark this answers is not in the document]
Comment (benoit-bremaud) [reply packages/api/src/feedback/feedback.module.ts:17]: Good catch (P1) — fixed in 25b4ab7: `Feedback` is now registered in `ormEntities` (typeorm.config.ts). Added an e2e test that hits the real DataSource so this regression is caught from now on.

File: packages/api/src/app.module.ts [hunk 2]
```diff
@@ -85,6 +86,9 @@ const bootstrapEnvironment = buildBootstrapEnvironmentConfig();
     LabelModule,
     ScanModule,
 
+    // Feedback module - public POST /feedback endpoint (epic #1026, #1027)
+    FeedbackModule,
+
```
[the remark this answers is not in the document]
Comment (benoit-bremaud) [reply packages/api/src/app.module.ts:91]: Fixed in 25b4ab7 — `Feedback` added to the `ormEntities` list in typeorm.config.ts (autoLoadEntities stays off). Verified by a new e2e test exercising the real DataSource.

File: packages/api/src/feedback/dtos/create-feedback.dto.ts [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+  @ApiProperty({ example: 'Mozilla/5.0 (...)' })
+  @IsString()
+  @MaxLength(512)
+  userAgent: string;
+
+  @ApiProperty({ type: FeedbackViewportDto })
+  @ValidateNested()
+  @Type(() => FeedbackViewportDto)
+  viewport: FeedbackViewportDto;
+
```
[the remark this answers is not in the document]
Comment (benoit-bremaud) [reply packages/api/src/feedback/dtos/create-feedback.dto.ts:117]: Fixed in 25b4ab7 — `viewport` now carries `@IsObject()` + `@IsNotEmptyObject()` (plus the existing `@ValidateNested()`/`@Type`), so a missing/invalid viewport returns 400 instead of a 500.

File: packages/api/src/feedback/dtos/create-feedback.dto.ts [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+  @ApiProperty({
+    nullable: true,
+    example: 'broken-feature',
+    description:
+      'Sub-category; must match the chosen category, or null for "other"',
+  })
+  @IsValidSubCategory()
+  subCategory: string | null;
+
```
[the remark this answers is not in the document]
Comment (benoit-bremaud) [reply packages/api/src/feedback/dtos/create-feedback.dto.ts]: Fixed in 25b4ab7 — `subCategory?: string | null` + `@ApiPropertyOptional`. Kept `@IsValidSubCategory()` (no `@IsOptional()`) so the pairing rule still requires a value for non-`other` categories.

File: packages/api/src/feedback/dtos/create-feedback.dto.ts [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+  @IsString()
+  @MaxLength(2048)
+  url: string;
+
+  @ApiProperty({ nullable: true, example: 'https://brasse-bouillon.com/' })
+  @IsOptional()
+  @IsString()
+  @MaxLength(2048)
+  referrer: string | null;
+
```
[the remark this answers is not in the document]
Comment (benoit-bremaud) [reply packages/api/src/feedback/dtos/create-feedback.dto.ts]: Fixed in 25b4ab7 — `referrer?: string | null` + `@ApiPropertyOptional`, aligning the TS type and Swagger with the existing `@IsOptional()` behaviour.

File: packages/api/src/feedback/entities/feedback.entity.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  /** Client-set submission timestamp (distinct from `created_at`). */
+  @Column({ type: 'datetime' })
+  client_timestamp: Date;
+
+  /** Widget version that produced the payload. */
+  @Column({ type: 'varchar', length: 40 })
+  widget_version: string;
+
+  /** Scroll depth ratio in [0, 1] at submission time. */
+  @Column({ type: 'float' })
```
[the remark this answers is not in the document]
Comment (benoit-bremaud) [reply packages/api/src/feedback/entities/feedback.entity.ts]: Fixed in 25b4ab7 — `scroll_depth` is now `type: 'real'` in the entity (and migration), consistent with equipment-profile.orm.entity.ts.

File: packages/api/src/database/migrations/1795000000000-AddFeedbackTable.ts [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        "referrer" varchar(2048),
+        "user_agent" varchar(512) NOT NULL,
+        "viewport_w" integer NOT NULL,
+        "viewport_h" integer NOT NULL,
+        "locale" varchar(20) NOT NULL,
+        "client_timestamp" datetime NOT NULL,
+        "widget_version" varchar(40) NOT NULL,
+        "scroll_depth" float NOT NULL,
+        "session_id" varchar(128) NOT NULL,
+        "created_at" datetime NOT NULL DEFAULT (CURRENT_TIMESTAMP)
```
[the remark this answers is not in the document]
Comment (benoit-bremaud) [reply packages/api/src/database/migrations/1795000000000-AddFeedbackTable.ts]: Done in 25b4ab7 — added CHECK constraints in the migration: `CHK_feedback_category` (allowed top-level values) and `CHK_feedback_scroll_depth` (0..1).

File: packages/api/src/database/migrations/1795000000000-AddFeedbackTable.ts [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        "message" text NOT NULL,
+        "url" varchar(2048) NOT NULL,
+        "referrer" varchar(2048),
+        "user_agent" varchar(512) NOT NULL,
+        "viewport_w" integer NOT NULL,
+        "viewport_h" integer NOT NULL,
+        "locale" varchar(20) NOT NULL,
+        "client_timestamp" datetime NOT NULL,
+        "widget_version" varchar(40) NOT NULL,
+        "scroll_depth" float NOT NULL,
```
[the remark this answers is not in the document]
Comment (benoit-bremaud) [reply packages/api/src/database/migrations/1795000000000-AddFeedbackTable.ts]: Fixed in 25b4ab7 — migration now uses `real` for `scroll_depth` (matching the entity and the rest of the schema).

**Comment by benoit-bremaud** [comment]
## Review summary

**CI:** green (api, build, SonarCloud, security-audit, changes).
**Reviews addressed:** 8/8 inline comments (1 Codex P1 + 7 Copilot) replied and resolved.

Key change since review — commit `25b4ab7`:
- **P1 fix:** `Feedback` registered in `ormEntities` (runtime `EntityMetadataNotFoundError` would otherwise break `POST /feedback`).
- `viewport` presence validation; `subCategory`/`referrer` optional in TS + Swagger; `scroll_depth` → `real`; CHECK constraints on `category` + `scroll_depth`.
- Added `test/feedback.e2e-spec.ts` (real DataSource + 400 contracts) — would have caught the P1.

682 unit tests + 5 e2e green. Commits: `13c4946`, `25b4ab7`. Ready to merge.
